core/src/main/java/org/sonarsource/sonarlint/core/log/MessageFormat.java

```diff
@@ -0,0 +1,491 @@
+    while (escapeCounter > 0) {
+      buffer.append(ESCAPE_CHAR);
+      escapeCounter--;
+    }
+  }
+
+  /**
+   * Appends the argument at the specified argument index (or, if no such argument exists, the "{}" delimiter pair) to
+   * the specified result char array at the specified position and returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 25 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeArgOrDelimPair(final Object[] arguments, final int argCount, final int currentArgument,
+    final StringBuilder buffer) {
+    if (currentArgument < argCount) {
+      recursiveDeepToString(arguments[currentArgument], buffer, null);
+    } else {
+      writeDelimPair(buffer);
+    }
+  }
+
+  /**
+   * This method performs a deep toString of the given Object.
+   * Primitive arrays are converted using their respective Arrays.toString methods while
+   * special handling is implemented for "container types", i.e. Object[], Map and Collection because those could
+   * contain themselves.
+   * <p>
+   * dejaVu is used in case of those container types to prevent an endless recursion.
+   * </p>
+   * <p>
+   * It should be noted that neither AbstractMap.toString() nor AbstractCollection.toString() implement such a
+   * behavior.
+   * They only check if the container is directly contained in itself, but not if a contained container contains the
+   * original one. Because of that, Arrays.toString(Object[]) isn't safe either.
+   * Confusing? Just read the last paragraph again and check the respective toString() implementation.
+   * </p>
+   * <p>
+   * This means, in effect, that logging would produce a usable output even if an ordinary System.out.println(o)
+   * would produce a relatively hard-to-debug StackOverflowError.
+   * </p>
+   *
+   * @param o      the Object to convert into a String
+   * @param str    the StringBuilder that o will be appended to
+   * @param dejaVu a list of container identities that were already used.
+   */
+  private static void recursiveDeepToString(@Nullable final Object o, final StringBuilder str, final Set<String> dejaVu) {
+    if (appendSpecialTypes(o, str)) {
+      return;
+    }
+    if (isMaybeRecursive(o)) {
+      appendPotentiallyRecursiveValue(o, str, dejaVu);
+    } else {
+      tryObjectToString(o, str);
+    }
+  }
+
+  private static boolean appendSpecialTypes(final Object o, final StringBuilder str) {
+    return appendSpecificTypes(str, o) || appendDate(o, str);
+  }
+
+  private static boolean appendSpecificTypes(final StringBuilder stringBuilder, @Nullable final Object obj) {
+    if (obj == null || obj instanceof String) {
+      stringBuilder.append((String) obj);
+    } else if (obj instanceof CharSequence) {
+      stringBuilder.append((CharSequence) obj);
+    } else if (obj instanceof Integer) { // LOG4J2-1437 unbox auto-boxed primitives to avoid calling toString()
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Move this trailing comment on the previous empty line. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)